data-raw/simulate-data.R

```diff
@@ -0,0 +1,243 @@
+  x |>
+    stringr::str_trunc(width = width, side = "left", ellipsis = "") |>
+    stringr::str_pad(width = width, side = "left", pad = "0")
+}
+
+create_fake_icd <- function(n, date = NULL) {
+  if (is.null(date)) {
+    date <- sample(c("1993-01-01", "1995-01-01"), 1)
+  }
+
+  if (date >= "1994-01-01") {
+    create_fake_icd10(n)
+  } else {
+    create_fake_icd8(n)
+  }
+}
+
+create_fake_icd8 <- function(n) {
+  here("data-raw/icd8-codes.txt") |>
+    read_lines() |>
+    str_trim() |>
+    as_tibble() |>
+    separate_wider_delim(
+      value,
+      delim = regex("   +"),
+      names = c("icd8", "description", "unknown"),
+      too_many = "merge"
+    ) |>
+    mutate(across(everything(), str_trim)) |>
+    mutate(
+      description = str_remove(description, "\\d+"),
+      icd8 = str_remove(icd8, "dia")
+    ) |>
+    pull(icd8) |>
+    sample(size = n, replace = TRUE)
+}
+
+create_fake_icd10 <- function(n) {
+  # from: https://medinfo.dk/sks/brows.php?s_nod=6308
+  here("data-raw/icd10-codes.csv") |>
+    read_csv2(show_col_types = FALSE) |>
+    pull(icd10) |>
+    sample(size = n, replace = TRUE)
+}
+
+create_fake_atc <- function(n) {
+  codeCollection::ATCKoodit |>
+    tibble::as_tibble() |>
+    dplyr::filter(stringr::str_length(Koodi) == 7) |>
+    dplyr::pull(Koodi) |>
+    sample(n, replace = TRUE)
+}
+
+create_fake_date <- function(n, from = "1977-01-01", to = lubridate::today()) {
+  seq(as_date(from), as_date(to), by = "day") |>
+    sample(n, replace = TRUE)
+}
+
+create_padded_integer <- function(n, length) {
+  purrr::map(1:length, \(ignore) sample(0:9, n, replace = TRUE)) |>
+    purrr::reduce(\(integer1, integer2) paste(integer1, integer2, sep = "")) |>
+    pad_integers(width = length)
+}
+
+create_fake_npu <- function(n) {
+  stringr::str_c(
+    "NPU",
+    create_padded_integer(n, 8)
+  )
+}
+
+create_fake_hovedspeciale_ans <- function(n) {
+  "https://www.dst.dk/da/Statistik/dokumentation/Times/forebyggelsesregistret/spec" |>
+    read_html() |>
+    html_element("table") |>
+    html_table() |>
+    pull(Tekst) |>
+    sample(n, replace = TRUE)
+}
+
+create_fake_drug_name <- function(atc) {
+  codeCollection::ATCKoodit |>
+    tibble::as_tibble() |>
+    dplyr::filter(Koodi %in% atc) |>
+    dplyr::pull(en) |>
+    sample(length(atc), replace = TRUE)
+}
+
+to_wwyy <- function(x) {
+  format(lubridate::as_date(x), format = "%W%y")
+}
+
+to_yyyymmdd <- function(x) {
+  format(lubridate::as_date(x), format = "%Y%m%d")
+}
+
+# Insert extra values to overrepresent certain values ------------------------------------------------------
+
+insertion_rate <- function(proportion) {
+  runif(1) < proportion
+}
+
+insert_specific_atc <- function(data, proportion = 0.3) {
+  # c(
+  #       "A10AB",
+  #       "A10AC",
+  #       "A10AD",
+  #       "A10AE",
+  #       "A10BA",
+  #       "A10BB",
+  #       "A10BD",
+  #       "A10BG",
+  #       "A10BH",
+  #       "A10BJ",
+  #       "A10BK",
+  #       "A10BX"
```

Comment:
FYI: these 11 ATC code levels contain all glucose-lowering drugs (by suffixing 01:99 you get to the individual drug level), e.g. A10BA is biguanides (a group of drugs) and A10BA02 is metformin (a specific type of biguanide).

The original test data script [lines 107-122 here](https://github.com/steno-aarhus/osdc/blob/387454a2acf273bf068b931dc3325d8828826b22/data-raw/testdata.R) manually added a random sample of these ATC codes to over-represent glucose-lowering drugs in the data set and cover a few edge cases of glucose-lowering drug types that are treated differently from the rest of the glucose-lowering drugs.
I'll list them here (off the top of my head) for future reference:

- A10BA02, metformin: censored in young women
- A10AE56, mix of insulin + liraglutid: is treated as a T2D-medication although it falls under the ATC codes for insulin.
- A10BJ02, liraglutid: is not used if sold under brand name "Saxenda" (for treatment of obesity, not diabetes)
- A10BJ06, semaglutid: no longer used for inclusions (it is increasingly used off-label for weightloss, not just for diabetes)
- A10BK01 + A10BK03, dapagliflozin + empagliflozin: no longer used for inclusions (are now also used for heart failure/kidney failure without diabetes).